org.scala-ide.sbt.core/src/org/scalaide/sbt/ui/internal/editor/SbtCompilationUnit.scala

```diff
@@ -0,0 +1,133 @@
+package org.scalaide.ui.internal.editor.sbt
+
+import scala.reflect.internal.util.BatchSourceFile
+import scala.reflect.io.AbstractFile
+
+import org.eclipse.core.resources.IFile
+import org.eclipse.core.resources.ResourcesPlugin
+import org.eclipse.core.runtime.IPath
+import org.eclipse.jface.text.IDocument
+import org.eclipse.ui.IEditorInput
+import org.eclipse.ui.part.FileEditorInput
+import org.scalaide.core.IScalaPlugin
+import org.scalaide.core.compiler.IPositionInformation
+import org.scalaide.core.compiler.ISourceMap
+import org.scalaide.core.compiler.InteractiveCompilationUnit
+import org.scalaide.core.extensions.SourceFileProvider
+import org.scalaide.core.resources.EclipseResource
+
+class SbtSourceFileProvider extends SourceFileProvider {
+  override def createFrom(path: IPath): Option[InteractiveCompilationUnit] = {
+    val root = ResourcesPlugin.getWorkspace().getRoot()
+    val file = Option(root.getFile(path))
+
+    file map (new SbtCompilationUnit(_))
+  }
+}
+
+object SbtCompilationUnit {
+  def fromEditor(scriptEditor: SbtEditor): SbtCompilationUnit = {
+    val input = scriptEditor.getEditorInput
+    if (input == null) throw new NullPointerException(s"No editor input for editor $scriptEditor. Hint: Maybe the editor isn't yet fully initialized?")
+    else {
+      val doc = scriptEditor.getDocumentProvider().getDocument(input)
+      fromEditorInput(input, doc)
+    }
+  }
+
+  private def getFile(editorInput: IEditorInput): IFile =
+    editorInput match {
+      case fileEditorInput: FileEditorInput if fileEditorInput.getName.endsWith("sbt") =>
+        fileEditorInput.getFile
+      case _ =>
+        throw new IllegalArgumentException(s"Editor input of file `${editorInput.getName}` is not a sbt file.")
+    }
+
+  private def fromEditorInput(editorInput: IEditorInput, doc: IDocument): SbtCompilationUnit =
+    new SbtCompilationUnit(getFile(editorInput), Some(doc))
+}
+
+class SbtSourceInfo(file: AbstractFile, override val originalSource: Array[Char]) extends ISourceMap {
+  def baseImports = "import sbt._, Keys._, dsl._"
+
+  val prefix = s"""|$baseImports
+                   |object $$container {
+                   |  def $$meth {
+                   |""".stripMargin
+
+  override val scalaSource =
+    s"""|$prefix
+        |${originalSource.mkString("")} }}""".stripMargin.toCharArray()
+
+  override lazy val sourceFile = new BatchSourceFile(file, scalaSource)
+
+  override val scalaPos = ScalaPosition
+  override val originalPos = OriginalPosition
+
+  override def scalaLine(line: Int): Int = line + 3
```

Comment:
Maybe make this magic constant be the number of lines in `$prefix`, since any change in the prefix would break this code.